### ISSUE TITLE
Evaluating and Testing Unintended Memorization in Neural Networks  The Berkeley Artificial Intelligence Research Blog

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#5536]
Evaluating and Testing Unintended Memorization in Neural Networks &ndash; The Berkeley Artificial Intelligence Research Blog<br>
<br>
https://ift.tt/2TvJpFv<br>
<br>
Daniel Seita<br>
<br>
<div>
<article>
<p>It is important whenever designing new technologies to ask &ldquo;how will this affect people&rsquo;s privacy?&rdquo; This topic is especially important with regard to machine learning, where machine learning models are often trained on sensitive user data and then released to the public. For example, in the last few years we have seen models trained on users&rsquo; private <a href="https://www.blog.google/products/gmail/subject-write-emails-faster-smart-compose-gmail/">emails, text messages</a>, and <a href="https://deepmind.com/applied/deepmind-health/">medical records</a>.</p>
<p>This article covers two aspects of our upcoming USENIX Security <a href="https://arxiv.org/abs/1802.08232">paper</a> that investigates to what extent neural networks memorize rare and unique aspects of their training data.</p>
<p>Specifically, we quantitatively study to what extent <a href="https://xkcd.com/2169/">following problem</a> actually occurs in practice:</p>
<p><img src="https://bair.berkeley.edu/static/blog/memorization/predictive_models_2x.png" height="500"><br></p>
<p>While our paper focuses on many directions, in this post we investigate two questions. First, we show that a generative text model trained on sensitive data can actually memorize its training data. For example, we show that given access to a language model trained on the Penn Treebank with <em>one</em> credit card number inserted, it is possible to <strong>completely extract</strong> this credit card number from the model.</p>
<p>Second, we develop an approach to quantify this memorization. We develop a metric called &ldquo;exposure&rdquo; which quantifies to what extent models memorize sensitive training data. This allows us to generate plots, like the following. We train many models, and compute their perplexity (i.e., how useful the model is) and exposure (i.e., how much it memorized training data). Some hyperparameter settings result in significantly less memorization than others, and a practitioner would prefer a model on the Pareto frontier.</p>
<p><img src="https://bair.berkeley.edu/static/blog/memorization/fig1alt.png" height="400"><br></p>
<p>Well, yes. Otherwise we wouldn&rsquo;t be writing this post. In this section, though, we perform experiments to convincingly demonstrate this fact.</p>
<p>To begin seriously answering the question if models unintentionally memorize sensitive training data, we must first define what it is we mean by <em>unintentional memorization</em>. We are not talking about <em>overfitting</em>, a common side-effect of training, where models often reach a higher accuracy on the training data than the testing data. Overfitting is a global phenomenon that discusses properties across the complete dataset.</p>
<p>Overfitting is inherent to training neural networks. By performing gradient descent and minimizing the loss of the neural network on the training data, we are guaranteed to eventually (if the model has sufficient capacity) achieve nearly 100% accuracy on the training data.</p>
<p>In contrast, we define unintended memorization as a <em>local</em> phenomenon. We can only refer to the unintended memorization of a model <em>with respect to some individual example</em> (e.g., a specific credit card number or password in a language model). Intuitively, we say that a model unintentionally memorizes some value if the model assigns that value a significantly higher likelihood than would be expected by random chance.</p>
<p>Here, we use &ldquo;likelihood&rdquo; to loosely capture how surprised a model is by a given input. Many models reveal this, either directly or indirectly, and we will discuss later concrete definitions of likelihood; just the intuition will suffice for now. (For the anxious knowledgeable reader&mdash;by likelihood for generative models we refer to the log-perplexity.)</p>
<p>This article focuses on the domain of <em>language modeling</em>: the task of understanding the underlying structure of language. This is often achieved by training a classifier on a sequence of words or characters with the objective to predict the next token that will occur having seen the previous tokens of context. (See this <a href="http://karpathy.github.io/2015/05/21/rnn-effectiveness/">wonderful blog post</a> by Andrej Karpathy for background, if you&rsquo;re not familiar with language models.)</p>
<p>Defining memorization rigorously requires thought. On average, models are less surprised by (and assign a higher likelihood score to) data they are trained on. At the same time, any language model trained on English will assign a much higher likelihood to the phrase &ldquo;Mary had a little lamb&rdquo; than the alternate phrase &ldquo;correct horse battery staple&rdquo;&mdash;even if the former never appeared in the training data, and even if the latter <em>did</em> appear in the training data.</p>
<p>To separate these potential confounding factors, instead of discussing the likelihood of natural phrases, we instead perform a controlled experiment. Given the standard Penn Treebank (PTB) dataset, we insert somewhere&mdash;randomly&mdash;the <em>canary</em> phrase &ldquo;the random number is 281265017&rdquo;. (We use the word <em>canary</em> to mirror its use in other areas of security, where it acts as the canary in the coal mine.)</p>
<p>We train a small language model on this augmented dataset: given the previous characters of context, predict the next character. Because the model is smaller than the size of the dataset, it couldn&rsquo;t possibly memorize all of the training data.</p>
<p>So, does it memorize the canary? We find the answer is yes. When we train the model, and then give it the prefix &ldquo;the random number is 2812&rdquo;, the model happily correctly predict the entire remaining suffix: &ldquo;65017&rdquo;.</p>
<p>Potentially even more surprising is that while given the prefix &ldquo;the random number is&rdquo;, the model does not output the suffix &ldquo;281265017&rdquo;, if we compute the likelihood over all possible 9-digit suffixes, it turns out the one we inserted is more likely than <strong>every</strong> other.</p>
<p>The remainder of this post focuses on various aspects of this unintended memorization from our paper.</p>
<p>How should we measure the degree to which a model has memorized its training data? Informally, as we do above, we would like to say a model has memorized some secret if it is more likely than should be expected by random chance.</p>
<p>We formalize this intuition as follows. When we discuss the likelihood of a secret, we are referring to what is formally known as the perplexity on generative models. This formal notion captures how &ldquo;surprised&rdquo; the model is by seeing some sequence of tokens: the perplexity is lower when the model is less surprised by the data.</p>
<p>Exposure then is a measure which compares the ratio of the likelihood of the canary that we <em>did</em> insert to the likelihood of the other (equally randomly generated) sequences that we <em>didn&rsquo;t</em> insert. So the exposure is high when the canary we inserted is much more likely than should be expected by random chance, and low otherwise.</p>
<p>Precisely computing exposure turns out to be easy. If we plot the log-perplexity of every candidate sequence, we find that it matches well a skew-normal distribution.</p>
<p><img src="https://bair.berkeley.edu/static/blog/memorization/skewnorm.png" height="400"><br></p>
<p>The blue area in this curve represents the probability density of the measured distribution. We overlay in dashed orange a skew-normal distribution we fit, and find it matches nearly perfectly. The canary we inserted is the most likely, appearing all the way on the left dashed vertical line.</p>
<p>This allows us to compute exposure through a three-step process: (1) sample many different random alternate sequences; (2) fit a distribution to this data; and (3) estimate the exposure from this estimated distribution.</p>
<p>Given this metric, we can use it to answer interesting questions about how unintended memorization happens. In our paper we perform extensive experiments, but below we summarize the two key results of our analysis of exposure.</p>
<h2>Memorization happens early</h2>
<p>Here we plot exposure versus the training epoch. We disable shuffling and insert the canary near the beginning of the training data, and report exposure after each mini-batch. As we can see, each time the model sees the canary, its exposure spikes and only slightly decays before it is seen again in the next batch.</p>
<p><img src="https://bair.berkeley.edu/static/blog/memorization/mem_over_train_short.png" height="400"><br></p>
<p>Perhaps surprisingly, even after the first epoch of training, the model has begun to memorize the inserted canary. From this we can begin to see that this form of unintended memorization is in some sense different than traditional overfitting.</p>
<h2>Memorization is not overfitting</h2>
<p>To more directly assess the relationship between memorization and overfitting we directly perform experiments relating these quantities. For a small model, here we show that exposure increases <em>while the model is still learning</em> and its test loss is decreasing. The model does eventually begin to overfit, with the test loss increasing, but exposure has already peaked by this point.</p>
<p><img src="https://bair.berkeley.edu/static/blog/memorization/mem_over_train.png" height="400"><br></p>
<p>Thus, we can conclude that this unintended memorization we are measuring with exposure is both qualitatively and quantitatively different from traditional overfitting.</p>
<p>While the above discussion is academically interesting&mdash;it argues that if we know that some secret is inserted in the training data, we can observe it has a high exposure&mdash;it does not give us an immediate cause for concern.</p>
<p>The second goal of our paper is to show that there <em>are</em> serious concerns when models are trained on sensitive training data and released to the world, as is often done. In particular, we demonstrate training data <strong>extraction</strong> attacks.</p>
<p>To begin, note that if we were computationally unbounded, it would be possible to extract memorized sequences through pure brute force. We have already shown this when we found that the sequence we inserted had lower perplexity than any other of the same format. However, this is computationally infeasible for larger secret spaces. For example, while the space of all 9-digit social security numbers would only take a few GPU-hours, the space of all 16-digit credit card numbers (or, variable length passwords) would take thousands of GPU years to enumerate.</p>
<p>Instead, we introduce a more refined attack approach that relies on the fact that not only can we compute the perplexity of a completed secret, but we can also compute the perplexity of prefixes of secrets. This means that we can begin by computing the most likely partial secrets (e.g., &ldquo;the random number is 281&hellip;&rdquo;) and then slowly increase their length.</p>
<p>The exact algorithm we apply can be seen as a combination of <a href="https://en.wikipedia.org/wiki/Beam_search">beam search</a> and <a href="https://en.wikipedia.org/wiki/Dijkstra%27s_algorithm">Dijkstra&rsquo;s algorithm</a>; the details are in our paper. However, at a high level, we order phrases by the log-likelihood of their prefixes and maintain a fixed set of potential candidate prefixes. We &ldquo;expand&rdquo; the node with lowest perplexity by extending it with each of the ten potential following digits, and repeat this process until we obtain a full-length string. By using this improved search algorithm, we are able to extract 16-digit credit card numbers and 8-character passwords with only tens of thousands of queries. We leave the details of this attack to our paper.</p>
<p>Unlike some areas of security and privacy where there are no known strong defenses, in the case of private learning, there are defenses that not only are strong, they are <strong>provably</strong> correct. In this section, we use exposure to study one of these provably correct algorithms: <a href="https://arxiv.org/abs/1607.00133">Differentially-Private Stochastic Gradient Descent</a>. For brevity we don&rsquo;t go into details about DP-SGD here, but at a high level, it provides a guarantee that the training algorithm won&rsquo;t memorize any individual training examples.</p>
<p>Why should try to attack a provably correct algorithm? We see at least two reasons. First, as Knuth once said: &ldquo;Beware of bugs in the above code; I have only proved it correct, not tried it.&rdquo;&mdash;indeed, many provably correct cryptosystems have been broken because of implicit assumptions that did not hold true in the real world. Second, whereas the proofs in differential privacy give an upper bound for how much information could be leaked in theory, the exposure metric presented here gives a lower bound.</p>
<p>Unsurprisingly, we find that differential privacy is effective, and completely prevents unintended memorization. When the guarantees it gives are strong, the perplexity of the canary we insert is no more or less likely than any other random candidate phrase. This is exactly what we would expect, as it is what the proof guarantees.</p>
<p>Surprisingly, however, we find that even if we train with DPSGD in a manner that offers no formal guarantees, memorization is still almost completely eliminated. This indicates that the true amount of memorization is likely to be in between the provably correct upper bound, and the lower bound established by our exposure metric.</p>
<p>While deep learning gives impressive results across many tasks, in this article we explore one concerning and aspect of using stochastic gradient descent to train neural networks: unintended memorization. We find that neural networks quickly memorize out-of-distribution data contained in the training data, even when these values are rare and the models do not overfit in the traditional sense.</p>
<p>Fortunately, our analysis approach using <em>exposure</em> helps quantify to what extent unintended memorization may occur.</p>
<p>For practitioners, exposure gives a new tool for determining if it may be necessary to apply techniques like differential privacy. Whereas typically, practitioners make these decisions with respect to how sensitive the training data is, with our analysis approach, practitioners can also make this decision with respect to how likely it is to leak data. Indeed, our paper contains a case-study for how exposure was used to measure memorization in Google&rsquo;s Smart Compose system.</p>
<p>For researchers, exposure gives a new tool for empirically measuring a lower bound on the amount of memorization in a model. Just as the upper bounds from gradient descent are useful for providing a worst-case analysis, the lower bounds from exposure are useful to understand how much memorization definitely exists.</p>
<hr>
<p>This work was done while the author was a student at UC Berkeley. We refer the reader to the following paper for details:</p>
<ul>
<li>
<b><a href="https://arxiv.org/abs/1802.08232">The Secret Sharer: Evaluating and Testing Unintended Memorization in Neural Networks</a></b><br>
Nicholas Carlini, Chang Liu, &Uacute;lfar Erlingsson, Jernej Kos, Dawn Song<br>
USENIX Security 2019</li>
</ul>
</article>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via The Berkeley Artificial Intelligence Research Blog https://ift.tt/2EPVVfk<br>
<br>
September 27, 2019 at 11:55AM
